text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+- Feature Name: variadic_tuples
+- Start Date: 2019-08-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+Tuples types are ordered set of type, but users can only use tuple with fixed arity.
+
+This RFC aims to allow the use of a _variadic tuple_ which is a tuple with an arbitrary arity.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Arbitrary tuple arity support
+
+Currently, when a user wants to either use or add behavior to tuples, he writes an impl for each tuple arity.
+For easier maintenance, it is usually done with a `macro_rules` and implemented up to 12 arity tuple. (see [ `Hash` implementation in `std`](https://github.com/rust-lang/rust/blob/master/src/libcore/hash/mod.rs)).
+
+Variadic tuple will provide several benefits considering trait implementation for tuple or using variadic tuples:
+
+- Implementations will be easier to write
+- Implementations will be easier to read and maintain
+- The compiler will compile implementation only for required tuple arity
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Let's call a _variadic tuple type_ a tuple type with an arbitrary arity and a _variadic tuple_ an instance of a variadic tuple type.
+
+A variadic tuple type is declared with `(..T)` and a variadic tuple type can be expanded with `(..Vec<T>)`.
```

Comment:
Here, "expanded" is really confusing to me. Really, what we're doing is putting constraints on what `T` actually is.

I'd just say that `T` represents *every* type in the tuple, and that we can replace `T` with something else to further restrict what's in the tuple. For example, `T` represents any type, `Vec<T>` represents vectors of any type, `usize` represents just `usize`, etc. More examples is generally good here.

Reply:
I guess this is not clear, because `(..Vec<T>)` is not a variadic tuple declaration. Only `(..T)` is allowed.

Maybe I can rephrase it with:

> A variadic tuple type is declared with `(..T)`. You can define other variadic tuples types from the idenfier `T`, for instance `(..Vec<T>)` defines a variadic tuple type those members are `Vec`s.
> ```rust
> Example:
> type TupleOfVec<(..T)> = (..Vec<T>);
> TupleOfVec<(uint, bool)> // (Vec<uint>, Vec<bool>)
> ```